sorting/pigeonhole_sort.cpp

```diff
@@ -0,0 +1,122 @@
+/**
+ * @file
+ * @brief Implementation of [Pigeonhole Sort algorithm]
+ * (https://en.wikipedia.org/wiki/Pigeonhole_sort)
+ * @author [Lownish](https://github.com/Lownish)
+ * @details
+ * Pigeonhole sorting is a sorting algorithm that is suitable for sorting lists
+ * of elements where the number of elements and the number of possible key values
+ * are approximately the same. It requires O(n + Range) time where n is number of
+ * elements in input array and ‘Range’ is number of possible values in array.
+ *
+ * The time Complexity of the algorithm is \f$O(n+N)\f$.
+ */
+
+#include <algorithm>  //for std::is_sorted
+#include <array>      //for std::array
+#include <cassert>    //for assert
+#include <iostream>   //for io operations
+
+/**
+ * @namespace sorting
+ * @brief Sorting algorithms
+ */
+namespace sorting {
+    /**
+     * @namespace pigonhole
+     * @brief Functions for [Pigeonhole Sort](https://en.wikipedia.org/wiki/Pigeonhole_sort) algorithm
+     */
+    namespace pigonhole {
+/**
+ * Pigeonhole sorting of array with array size n
+ * The function will sort the array through Pigeonhole algorithm and print
+ * @param unsorted array of elements
+ * @returns none
+ */
+template <std::size_t N>
+void pigeonSort(std::array<int, N> arr, int n) {
+    // Finding min and max*
+    int min = arr.at(0);
+    int max = arr.at(0);
+
+    for (int i = 0; i < n; i++) {
+        if (min > arr[i]) {
+            min = arr[i];
+        }
+        if (max < arr[i]) {
+            max = arr[i];
+        }
+    }
+
+    // Range refers to the number of holes required
+    int range = max - min + 1;
+    int* hole = new int[range]();
```

Comment:
Memory is not being freed and may cause memory leaks.
Free up the allocated memory using `delete`.